main/lsp/LSPPreprocessor.cc

```diff
@@ -86,32 +88,35 @@ void LSPPreprocessor::mergeFileChanges(absl::Mutex &mtx, QueueState &state) {
     // Check if we should cancel the slow path.
     const auto &gs = ttgs.getGlobalState();
     const auto runningSlowPath = gs.getRunningSlowPath();
-    // Only try to cancel if the cancelation feature is enabled.
+    // Only try to cancel if the cancelation feature is enabled AND the typechecking thread is running the slow path.
     if (config.opts.lspCancelableSlowPathEnabled && runningSlowPath.has_value()) {
         const auto &[committed, end] = runningSlowPath.value();
-        // Prune history for all messages no longer in queue or being processed by typechecking thread.
-        ttgs.pruneBefore(min(committed, earliestActiveEditVersion));
-        for (auto it = pendingRequests.begin(); it != pendingRequests.end(); it++) {
-            const auto &msg = *it;
-            if (msg->isNotification() && msg->method() == LSPMethod::SorbetWorkspaceEdit) {
-                Timer timeit(logger, "tryCancelSlowPath");
-                auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
-                auto combinedUpdates = ttgs.getCombinedUpdates(committed + 1, params->updates.versionEnd);
-                if (combinedUpdates.canTakeFastPath && gs.tryCancelSlowPath(params->updates.versionEnd)) {
-                    logger->debug("[Preprocessor] Canceling typechecking, as edits {} thru {} can take fast path.",
-                                  params->updates.versionStart, params->updates.versionEnd);
-                    params->updates = move(combinedUpdates);
+        earliestActiveEditVersion = min(committed, earliestActiveEditVersion);
+
+        // Avoid canceling if the currently-running slow path has already been canceled.
+        if (!gs.wasTypecheckingCanceled()) {
+            for (auto it = pendingRequests.begin(); it != pendingRequests.end(); it++) {
+                const auto &msg = *it;
+                if (msg->isNotification() && msg->method() == LSPMethod::SorbetWorkspaceEdit) {
+                    Timer timeit(logger, "tryCancelSlowPath");
+                    auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
+                    auto combinedUpdates = ttgs.getCombinedUpdates(committed + 1, params->updates.versionEnd);
```

Comment:
This line calls indexer to re-create the index trees, which is why it's somewhat expensive.

Reply:
could you leave this comment in code?